README.md
PDF-Marking-Tool
================

A solution for PDF-based on-screen marking of assignments, tests, etc. Used by a number of universities in South Africa

This project was forked from the "Unisa Tool for Onscreen Marking". This fork will retain the purpose of the original UTOM, but may choose to add additional capabilities to make the solution applicable to a wider user base.

The project is available under the GPL v2 licence.

Please ensure that all distributions of the code clearly acknowledges the role of the University of South Africa, ideally with a link to the project's original home at http://www.unisa.ac.za/Default.asp?Cmd=ViewContent&ContentID=23049

UNISA eMarking Suite is delivered to UNISA under the FLOSS licencing regime

## Installation
- Ensure Adobe Acrobat DC is installed. This is a prerequisite for the tool
- Install the marking tool using the provided installer (You will require administrative rights on your workstation)

## Usage Notes
- After installation, open any PDF document, but right-clicking on the file and selecting "Open with Adobe Acrobat DC"
- Note: Only one PDF document should be opened at a time, since the marking tool references the current document
- After the document is opened, select "Tools" > "Add-on Tools". This will open the marking tools
- Apply marking tools as needed
- Once marking is complete, use the "Count Marks" tool. This will add a new "Results" page to the document with a breakdown of the marks applied
- On the "Results" page, click the "Finalize" button, to complete the document and lock it for further marks.



RELEASE-NOTES.md
# RELEASE NOTES

## v1.8.4

Date Released: 2021-02-18
Developer: Greg Fullard

### Changes
#### Updated wording of the Finalise button
Updated the popup dialog wording in edge-case scenario where an assignment is counted, then saved and closed, then subsequently reopened and Finalised.

## v1.8.3

Date Released: 2021-02-03
Developer: Greg Fullard

### Changes
#### Recovered previously resolved defects
Due to incorrect baseline source code used at the start of v1.8 development, three defects that were previously fixed were introduced back during v1.8 development.

The defects are:
 - Inconsistent rounding of marks
 - Inconsistent naming of finalised file (if document is saved and closed before finalising)
 - Inconsistent annotation functionality in landscape mode

Previous defect fixes were added back to the main codebase and tested

## v1.8

Date Released: 2020-10-27
Developer: Greg Fullard
  
### Changes   
#### Updated Windows installation to work with Adobe Acrobat DC Pro on Windows 10
The installer script was refactored to work on Acrobat DC Pro on Windows 10. Key changes to make this possible included:
- Configuring the installer to run with administrator privileges
- Updating the way in which the installer verifies the current version of Acrobat
- Refactored and added comments to the installer in an effort to make the code easier to work with in the future

#### Improved the default startup configuration of marker plugin
Edge-case usability of the marking tool was very problematic due to a lack of error handling in the code. The result to an end-user was that the plugin simply wouldn't load and the user would have no clue why not.

I added various error checks in the config script so that the marking tool plugin loads nonetheless, but then provides the user with an indication of any errors that happened during startup. 

#### Resized tool icons
The default icon size used in earlier versions of the tool are larger that what is allowed by the app.addToolButton method of the Acrobat JavaScript API. Icons ware 100x100 pixels, but the documentation states that a maximum of 20x20 is allowed. The tool would intermittently fail to load due to the icons.

I added error handling code to give the end-user a heads-up if this is the case

Initially, I did resize the icons to 20x20, but then the icons were too small to user. Instead, I resized to 50.50 pixels. Even though this is still larger than the allowed size, it seems to work correctly

#### Added Developer Notes
Added a DEVELOPER_NOTES.MD file in the docs folder. This document contains various references and notes that were uncovered while working with the tool.

### Test cases
Earlier versions of the tool did not include any test cases. The following informal test scenarios were identified and tested:

* Installer
  * PASSED: Run installer on Windows 10 machine with neither Reader or Acrobat DC
  * PASSED: Run installer on Windows 10 machine with just Reader installed
  * PASSED: Run installer on Windows 10 machine with Acrobat DC installed
  * PASSED: Attempt to install with normal user (without admin privileges)
* Uninstaller
  * PASSED: Uninstall marking tool
* Plugin
  * PASSED: Open without any PDF document
  * PASSED: Open with PDF document
  * PASSED: Perform marking, count marks, finalise document

## Version 1.7 - February 2015

      Proudly Developed By
   
      KYLE BOWDEN in conjuction with Learning Curve
      kyle247365@gmail.com | +27 84 738 9643
   
      Enhancements:
        - Update installer to set Javascript prefs in Acrobat 11 & Higher
        - Fixed comments rolling off page
        - Added installer for Mac OSX
        - Minor refactors / tweaks

## Version 1.6 - 1.6.1 -  November 2012
  
      Proudly Developed By
   
      KYLE BOWDEN in conjuction with Learning Curve
      kyle247365@gmail.com | +27 84 738 9643
   
      Enhancements:
        - Minor system updates / tweaks
        - Ported application to the Windows 7 32-bit platform
   
## Version 0.1 - 1.5 - October 2009

      Proudly Developed By
 
      THE AESIR DEVELOPMENT SQUAD 
      www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico



src/MarkingTool/v1.6.1/src/config2.js
/* 
   UNISA eMarking Suite
   is delivered to UNISA under the
   FLOSS licencing regime

Version 1.6 - 1.6.1 -  November 2012
  
   Proudly Developed By
   KYLE BOWDEN in conjuction with Learning Curve
   kyle247365@gmail.com | +27 84 738 9643
   
   Enhancements:
    - Minor system updates / tweaks
    - Ported application to the Windows 7 32-bit platform
   
Version 0.1 - 1.5 - October 2009

   Proudly Developed By
 
   THE AESIR DEVELOPMENT SQUAD 
   www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico
*/

app.beginPriv();

var aActiveDocs = app.activeDocs;
var aNewDoc = aActiveDocs[0];

/*
Version 1.6 - 1.6.1 -  November 2012
  
   Proudly Developed By
   KYLE BOWDEN in conjuction with Learning Curve
   kyle247365@gmail.com | +27 84 738 9643
   
   Enhancements:
    - Minor system updates / tweaks
    - Ported application to the Windows 7 32-bit platform
   
Version 0.1 - 1.5 - October 2009

   Proudly Developed By
 
   THE AESIR DEVELOPMENT SQUAD 
   www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico
*/
try {
var numpages = aNewDoc.numPages;

for (var i=0; i < numpages; i++) {
aNewDoc.removeField("btn" + i);
}
} catch(Error) {}

var inUMSSession = "true";
try {
var edtFinish = aNewDoc.getField("edtFinish");
if(edtFinish.value == "FINAL_DONE") {
inUMSSession = "false";
}
} catch(Error) {}

/*
Version 1.6 - 1.6.1 -  November 2012
  
   Proudly Developed By
   KYLE BOWDEN in conjuction with Learning Curve
   kyle247365@gmail.com | +27 84 738 9643
   
   Enhancements:
    - Minor system updates / tweaks
    - Ported application to the Windows 7 32-bit platform
   
Version 0.1 - 1.5 - October 2009

   Proudly Developed By
 
   THE AESIR DEVELOPMENT SQUAD 
   www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico
*/
try {
var fileString = aNewDoc.path.substring(0, 4);

if(fileString == "http") {
inUMSSession = "false";
}
} catch(Error) {}

/*
Version 1.6 - 1.6.1 -  November 2012
  
   Proudly Developed By
   KYLE BOWDEN in conjuction with Learning Curve
   kyle247365@gmail.com | +27 84 738 9643
   
   Enhancements:
    - Minor system updates / tweaks
    - Ported application to the Windows 7 32-bit platform
   
Version 0.1 - 1.5 - October 2009

   Proudly Developed By
 
   THE AESIR DEVELOPMENT SQUAD 
   www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico
*/
if(inUMSSession == "true") {
var btnOpenRubric = aNewDoc.addField("btnOpenRubric", "button", 0, [5, 5, 70, 30]);
btnOpenRubric.buttonSetCaption("Open Rubric");
btnOpenRubric.setAction("MouseUp", "openRubricForMarking(aNewDoc)");
btnOpenRubric.borderStyle = "beveled";
btnOpenRubric.highlight = "push";
btnOpenRubric.lineWidth = 2;

app.hideToolbarButton("Hand");
app.hideToolbarButton("ZoomIn");
app.hideToolbarButton("Zoom100");
app.hideToolbarButton("PanAndZoom");
app.hideToolbarButton("Loupe");

aNewDoc.importIcon("addCommentMark", "/C/Program Files/UNISA/commentmark.png");
aNewDoc.importIcon("addMark", "/C/Program Files/UNISA/mark.png");
aNewDoc.importIcon("addTick", "/C/Program Files/UNISA/tickmark.png");
aNewDoc.importIcon("addCross", "/C/Program Files/UNISA/crossmark.png");
aNewDoc.importIcon("addCount", "/C/Program Files/UNISA/count.png");
aNewDoc.importIcon("addStamp", "/C/Program Files/UNISA/check.png");
aNewDoc.importIcon("addDeselect", "/C/Program Files/UNISA/deselect.png");
aNewDoc.importIcon("addHalfTick", "/C/Program Files/UNISA/halftickmark.png");

var addCommentMarkIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addCommentMark"));
var addMarkIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addMark"));
var addTickIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addTick"));
var addCrossIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addCross"));
var addCountIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addCount"));
var addStampIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addStamp"));
var addDeselectIcon = util.iconStreamFromIcon(aNewDoc.getIcon("addDeselect"));
var addHalfTick = util.iconStreamFromIcon(aNewDoc.getIcon("addHalfTick"));

/*
Version 1.6 - 1.6.1 -  November 2012
  
   Proudly Developed By
   KYLE BOWDEN in conjuction with Learning Curve
   kyle247365@gmail.com | +27 84 738 9643
   
   Enhancements:
    - Minor system updates / tweaks
    - Ported application to the Windows 7 32-bit platform
   
Version 0.1 - 1.5 - October 2009

   Proudly Developed By
 
   THE AESIR DEVELOPMENT SQUAD 
   www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico
*/
app.addToolButton
({
cName: "toolAddHalfTick",
oIcon: addHalfTick,
cExec: "addMark(aNewDoc, 'HALFT');",
cTooltext: "Add a Half Tick Mark.",
cMarked: "setIndentOnHalfTick();",
nPos: 0
});

app.addToolButton
({
cName: "toolAddTick",
oIcon: addTickIcon,
cExec: "addMark(aNewDoc, 'TICK');",
cTooltext: "Add a Tick.",
cMarked: "setIndentOnTick();",
nPos: 1
});

app.addToolButton
({
cName: "toolAddStamp",
oIcon: addStampIcon,
cExec: "addMark(aNewDoc, 'CHECK');",
cTooltext: "Add a Stamp.",
cMarked: "setIndentOnCheck();",
nPos: 2
});

app.addToolButton
({
cName: "toolAddCross",
oIcon: addCrossIcon,
cExec: "addMark(aNewDoc, 'CROSS');",
cTooltext: "Add a Cross.",
cMarked: "setIndentOnCross();",
nPos: 3
});

app.addToolButton
({
cName: "toolDeselect",
oIcon: addDeselectIcon,
cExec: "removeContinuesMarking(aNewDoc)",
cTooltext: "Deselect Mark Tool.",
cEnable: "setEnableOnDeselect();",
nPos: 4
});

app.addToolButton
({
cName: "toolAddMark",
oIcon: addMarkIcon,
cExec: "addMark(aNewDoc, 'MARK');",
cTooltext: "Add a Mark.",
cMarked: "setIndentOnMark();",
nPos: 5
});

app.addToolButton
({
cName: "toolAddCommentMark",
oIcon: addCommentMarkIcon,
cExec: "addMark(aNewDoc, 'COMMENTM');",
cTooltext: "Comment Mark.",
cMarked: "setIndentOnCommentMark();",
nPos: 6
});

app.addToolButton
({
cName: "toolAddCount",
oIcon: addCountIcon,
cExec: "countMarks(aNewDoc)",
cTooltext: "Count Marks.",
nPos: 7
});

app.addToolButton
({
cName: "toolVersion",
cExec: "",
cLabel: "v1.6",
cTooltext: "Click Here",
cExec: "showAboutInformation()",
nPos: 8
});
}

// ACCESS DOCUMENT PROPERTIES
for(var oo in aNewDoc) {
    (oo + ", " + aNewDoc[oo]);
}

app.endPriv();
/*
Version 1.6 - 1.6.1 -  November 2012
  
   Proudly Developed By
   KYLE BOWDEN in conjuction with Learning Curve
   kyle247365@gmail.com | +27 84 738 9643
   
   Enhancements:
    - Minor system updates / tweaks
    - Ported application to the Windows 7 32-bit platform
   
Version 0.1 - 1.5 - October 2009

   Proudly Developed By
 
   THE AESIR DEVELOPMENT SQUAD 
   www.aesir.co.za | info@aesir.co.za | +27 11 702 9000
 
      CREDITS
      Lead Developer    :: War Commander      :: Kyle Bowden
      Business Analyst  :: Field General      :: Willy Gadney
      Test Squad        :: Clean Out Crew     :: Herman Van Wyk & Tina Kanniah
      Dev Support       :: Backup             :: Nelson Baloyi
      LC Designer       :: Artillery          :: Lentswe Morule
      Installer         :: Mobilizer          :: Thabahla Shoko
      Enviro & Food     :: Crew Support       :: Khosi Gwala
      Architect         :: Special Operations :: Luigi D'Amico
*/